macOS logout/login doesn't restart cleanly

**Describe the bug**
Logging out and then logging in to user with espanso installed fails to restart cleanly.

**To Reproduce**
Steps to reproduce the behavior:
1. Reboot macOS.
2. Login to user with espanso installed. ps shows:
<img width="704" alt="Screenshot 2022-12-30 at 19 56 16" src="https://user-images.githubusercontent.com/82220/210108015-4490c55f-5b1a-4787-8a39-b5166eb53560.png">

3. Logout
4. Login again.
5. espanso displays the following modal:
<img width="633" alt="Screenshot 2022-12-30 at 19 57 25" src="https://user-images.githubusercontent.com/82220/210107905-ed328983-a3e9-419c-a7bb-4944f9ffb070.png">

6. pse shows:
<img width="607" alt="Screenshot 2022-12-30 at 19 57 45" src="https://user-images.githubusercontent.com/82220/210108049-89d07c4f-4819-4d64-9465-5db9804e11f9.png">

7. To clean things up, it's necessary to kill -9 the process and and reload espanso

**Expected behavior**
espanso should unload all process on logout and restart cleanly on login.

**Setup information**
- OS: macOS 12.6.2 (Monterey)
- Version: espanso 2.1.8

I see the same on:

- macOS 26.5 (Apple Silicon)
- espanso 2.3.0 (Homebrew cask managed via nix-darwin, if it matters)
- Login via standard logout/login (not fast user switching)

### Symptom

AFAICT, the launchd job `com.federicoterzi.espanso` fires at login, exits non-zero, and is not retried. `espanso status` reports "espanso is not running" and the plist's stderr captures:

```
[daemon] [ERROR] daemon is already running!

thread 'main' panicked at espanso/src/cli/launcher/mod.rs:196:49:
failed to launch daemon: unexpected error, 'espanso daemon' returned a
non-zero exit code.
```

### My (Current) Workaround

I created a thin shell wrapper that clears the stale state before exec'ing espanso. The plist points at the wrapper instead of espanso directly.

`~/.local/bin/espanso-launcher` (in my `$PATH`):

```bash
#!/usr/bin/env bash
set -eu

RUNTIME="$HOME/Library/Caches/espanso"
rm -f \
    "$RUNTIME/espanso-daemon.lock" \
    "$RUNTIME/espanso-worker.lock" \
    "$RUNTIME/espansodaemonv2.sock" \
    "$RUNTIME/espansoworkerv2.sock"

exec /Applications/Espanso.app/Contents/MacOS/espanso launcher
```

`~/Library/LaunchAgents/com.federicoterzi.espanso.plist` — only the `ProgramArguments` block changes:

```xml
<key>ProgramArguments</key>
<array>
  <string>/Users/myusername/.local/bin/espanso-launcher</string>
</array>
```

I manage this through a stow-based dotfiles repo so the wrapper and plist are version-controlled and survive reinstalls. It seems to be working so far, but I honestly can't tell you whether I've had an upgrade come through which might rewrite the plist file.

hashtag ymmv, of course.

I'll see if I can drop a patch in the coming days. Will need to figure out the dev path. I don't have much Rust experience.

--- a/espanso/src/cli/daemon/mod.rs
+++ b/espanso/src/cli/daemon/mod.rs
@@ -68,6 +68,14 @@ fn daemon_main(args: CliModuleArgs) -> i32 {
 
     // Make sure only one instance of the daemon is running
     let lock_file = acquire_daemon_lock(&paths.runtime);
+
+    // On macOS, a daemon from a previous login session can survive a
+    // logout/login (see GitHub issue #1492), keeping the daemon lock held while
+    // no longer serving the current session. In that case, try to take over
+    // from the stale daemon instead of giving up.
+    #[cfg(target_os = "macos")]
+    let lock_file = lock_file.or_else(|| reclaim_stale_daemon_lock(&paths.runtime));
+
     if lock_file.is_none() {
         error!("daemon is already running!");
         return DAEMON_ALREADY_RUNNING;
@@ -192,6 +200,50 @@ fn daemon_main(args: CliModuleArgs) -> i32 {
     }
 
     DAEMON_SUCCESS
+}
+
+// Take over the daemon lock from a stale daemon that survived a previous login
+// session (macOS, GitHub issue #1492). We first ask the existing daemon to
+// exit so it releases the lock cleanly (which also tears down its worker); if
+// it does not release within a short window (e.g. it is wedged), we clear the
+// stale lock file so a fresh daemon can acquire a new one.
+#[cfg(target_os = "macos")]
+fn reclaim_stale_daemon_lock(runtime_dir: &Path) -> Option<crate::lock::Lock> {
+    warn!("daemon lock is held, attempting to take over from a possibly stale daemon...");
+
+    // Ask the existing daemon to exit. A live, responsive daemon releases the
+    // lock (and shuts down its worker); a duplicate launch simply hands off.
+    match crate::ipc::create_ipc_client_to_daemon(runtime_dir) {
+        Ok(mut daemon_ipc) => {
+            if let Err(err) = daemon_ipc.send_async(IPCEvent::Exit) {
+                error!("unable to send termination signal to daemon process: {err}");
+            }
+        }
+        Err(err) => {
+            error!("could not establish IPC connection with daemon: {err}");
+        }
+    }
+
+    // Wait for the stale daemon to release the lock.
+    let now = Instant::now();
+    while now.elapsed() < std::time::Duration::from_secs(3) {
+        if let Some(lock) = acquire_daemon_lock(runtime_dir) {
+            return Some(lock);
+        }
+
+        std::thread::sleep(std::time::Duration::from_millis(200));
+    }
+
+    // The holder did not release the lock (e.g. an unresponsive orphan). Clear
+    // the stale lock file so a fresh daemon can acquire a new one; the orphan
+    // keeps its now-unlinked lock file and no longer blocks us.
+    warn!("stale daemon did not release the lock, clearing the stale lock file");
+    if let Err(err) = crate::lock::clear_daemon_lock(runtime_dir) {
+        error!("unable to clear stale daemon lock file: {err}");
+        return None;
+    }
+
+    acquire_daemon_lock(runtime_dir)
 }
 
 fn terminate_worker_if_already_running(runtime_dir: &Path) {

--- a/espanso/src/ipc.rs
+++ b/espanso/src/ipc.rs
@@ -54,6 +54,11 @@ pub fn create_ipc_client_to_worker(runtime_dir: &Path) -> Result<impl IPCClient<
     create_ipc_client(runtime_dir, "workerv2")
 }
 
+#[cfg(target_os = "macos")]
+pub fn create_ipc_client_to_daemon(runtime_dir: &Path) -> Result<impl IPCClient<IPCEvent>> {
+    create_ipc_client(runtime_dir, "daemonv2")
+}
+
 fn create_ipc_server(runtime_dir: &Path, name: &str) -> Result<impl IPCServer<IPCEvent>> {
     espanso_ipc::server(&format!("espanso{name}"), runtime_dir)
 }

--- a/espanso/src/lock.rs
+++ b/espanso/src/lock.rs
@@ -21,8 +21,15 @@ use anyhow::Result;
 use fs2::FileExt;
 use std::{
     fs::{File, OpenOptions},
-    path::Path,
+    path::{Path, PathBuf},
 };
+
+const DAEMON_LOCK_NAME: &str = "espanso-daemon";
+const WORKER_LOCK_NAME: &str = "espanso-worker";
+
+fn lock_file_path(runtime_dir: &Path, name: &str) -> PathBuf {
+    runtime_dir.join(format!("{name}.lock"))
+}
 
 pub struct Lock {
     lock_file: File,
@@ -36,7 +43,7 @@ impl Lock {
     }
 
     fn acquire(runtime_dir: &Path, name: &str) -> Option<Lock> {
-        let lock_file_path = runtime_dir.join(format!("{name}.lock"));
+        let lock_file_path = lock_file_path(runtime_dir, name);
         let lock_file = OpenOptions::new()
             .read(true)
             .write(true)
@@ -65,9 +72,47 @@ impl Drop for Lock {
 }
 
 pub fn acquire_daemon_lock(runtime_dir: &Path) -> Option<Lock> {
-    Lock::acquire(runtime_dir, "espanso-daemon")
+    Lock::acquire(runtime_dir, DAEMON_LOCK_NAME)
 }
 
 pub fn acquire_worker_lock(runtime_dir: &Path) -> Option<Lock> {
-    Lock::acquire(runtime_dir, "espanso-worker")
+    Lock::acquire(runtime_dir, WORKER_LOCK_NAME)
+}
+
+#[cfg(target_os = "macos")]
+pub fn clear_daemon_lock(runtime_dir: &Path) -> Result<()> {
+    let lock_file_path = lock_file_path(runtime_dir, DAEMON_LOCK_NAME);
+    if lock_file_path.exists() {
+        std::fs::remove_file(&lock_file_path)?;
+    }
+    Ok(())
+}
+
+#[cfg(all(test, target_os = "macos"))]
+mod tests {
+    use super::*;
+    use tempdir::TempDir;
+
+    #[test]
+    fn clear_daemon_lock_allows_takeover_while_stale_lock_is_held() {
+        let dir = TempDir::new("espansolock").unwrap();
+        let runtime = dir.path();
+
+        // An orphaned daemon acquires the lock and keeps holding it.
+        let orphan = acquire_daemon_lock(runtime);
+        assert!(orphan.is_some());
+
+        // A new daemon can't acquire the lock while the orphan holds it.
+        assert!(acquire_daemon_lock(runtime).is_none());
+
+        // Clearing the stale lock file lets the new daemon take over, even
+        // though the orphan is still holding its (now unlinked) lock file.
+        clear_daemon_lock(runtime).unwrap();
+        let takeover = acquire_daemon_lock(runtime);
+        assert!(takeover.is_some());
+
+        // Keep the orphan alive until the end so the assertion above proves
+        // the inode swap works, not that the orphan simply released.
+        drop(orphan);
+    }
 }

